[alfaloo] ip

alfaloo opened the pull request:
# Lulu Task Manager 🤖 

Lulu Task Manager is a simple command-line task manager written in Java. Keep track of your tasks with ease!

## Table of Contents

- [Introduction](#introduction)
- [Features](#features)
- [Installation](#installation)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)

## Introduction

Lulu Task Manager is a Java-based task management application. It allows users to add, mark, unmark, and delete tasks, as well as perform queries based on specific criteria. The application utilizes a command-line interface for a straightforward user experience.

## Features

- **Task Management:** Add, mark, unmark, and delete tasks.
- **Task Types:** Support for different task types, including Todo, Deadline, and Event.
- **Query Tasks:** Find tasks based on specific criteria like name or date.

- [x] Task 1
- [ ] Task 2
- [ ] Task 3

## Installation

Ensure you have Java installed on your system. Clone the repository and compile the Java files.

```bash
# Clone the repository
git clone https://github.com/your-username/lulu-task-manager.git

# Compile the Java files
javac -d . src/*.java
```

This is Lulu `Lulu chatbot = new Lulu();`.

```java
public static void main(String[] args) {
        Lulu chatbot = new Lulu();
        UI.start();
        try {
            chatbot.respond();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        UI.exit();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

alfaloo committed:
Level-0

alfaloo committed:
Level-0

alfaloo committed:
Level-0

alfaloo committed:
Level-0

alfaloo committed:
Level-0

alfaloo committed:
Level-0

alfaloo committed:
Level-0

alfaloo committed:
Level-1

alfaloo committed:
Level-2

alfaloo committed:
Level-3

alfaloo committed:
message

alfaloo committed:
testing

alfaloo committed:
Level-5

alfaloo committed:
Level-5

alfaloo committed:
Level-5

alfaloo committed:
Level-5

alfaloo committed:
Level-5

alfaloo committed:
A-Exceptions

alfaloo committed:
Level-5

alfaloo committed:
Level-5

alfaloo committed:
A-Exceptions

alfaloo committed:
Level-6

alfaloo committed:
A-Collections

alfaloo committed:
A-Enums

alfaloo committed:
implement save to file when adding new task

alfaloo committed:
implement save for mark, unmark, and delete

alfaloo committed:
minor changes

alfaloo committed:
merge branch-Level-7 with master

alfaloo committed:
add LocalDate

alfaloo committed:
merge branch-Level-8 with master

alfaloo committed:
add queryByTime feature

alfaloo committed:
made structure more oop

alfaloo committed:
separate code into packages

alfaloo committed:
merge gradle

alfaloo committed:
implement jUnit

alfaloo committed:
create jar

alfaloo committed:
add javadocs

alfaloo committed:
fix coding standard

alfaloo committed:
minor changes

alfaloo committed:
implement find feature

alfaloo committed:
merge with branch-A-JavaDoc

alfaloo committed:
merge with branch-A-CodingStandard

alfaloo committed:
merge with branch-Level-9

alfaloo committed:
minor change

alfaloo committed:
move greeting and farewell to UI class

alfaloo committed:
update readme file

hiivan commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * An abstract class representing a command in the task management system.
+ */
+public abstract class Command {
```
Good job abstracting all your command classes!

hiivan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,249 @@
+     * @return The ManageCommand object.
+     * @throws LuluException If an error occurs during parsing.
+     */
+    public Command manageTasks(Command.Types command, String input) throws LuluException {
```
Good use of enum data structure!

hiivan reviewed (commented):
Overall good job abstracting all your classes. Error handling is very robust too. Good job!

alfaloo committed:
add checkstyle

alfaloo committed:
update checkstyle

alfaloo committed:
minor changes

alfaloo committed:
minor changes

alfaloo committed:
javafx

Add JavaFX support to the project

Current situation: Project lacks a GUI framework,
limiting user experience.

Why it needs to change: Enhance UI and enable
interactive features with JavaFX.

What is being done: Introduce JavaFX dependencies
and configurations to the project.

Why it is done that way: Follow JavaFX best practices
for smooth integration.

Other info: Initial commit lays groundwork for UI
development and JavaFX feature integration.

alfaloo committed:
finish javafx gui tutorial

alfaloo committed:
integrate javafx for Lulu programme

alfaloo committed:
implement fxml

Add FXML support for modular UI development

Current: Project lacks FXML for JavaFX UI definition.

Why: FXML enables modular, readable UI development.

What: Introduce FXML files, update controllers.

Why this way: FXML aligns with JavaFX best practices.

Other: Lays foundation for more maintainable UI.

alfaloo committed:
minor changes

alfaloo committed:
add safe varargs

Improve varargs safety

Current situation: Unsafe varargs usage in some methods.

Why it needs to change: Risk of ClassCastException and
runtime issues.

What is being done: Implement safe varargs handling.

Why it is done that way: Explicit type checks for safer
usage.

Other info: Focus on method safety; future commits may
extend improvements.

ChuaZenKhoon commented:
Hi, just a minor detail regarding the commit message: Uppercase first letter :) You can refer to this site! You can also check the ip-progress dashboard if the git-standard tag is green!

[Git convention](https://se-education.org/guides/conventions/git.html)

ChuaZenKhoon reviewed (commented):
Hi, just a minor detail regarding the commit message: Uppercase first letter :) You can refer to this site! You can also check the ip-progress dashboard if the git-standard tag is green!

e.g Add safe varargs instead of add safe varags

[Git convention](https://se-education.org/guides/conventions/git.html)

ChuaZenKhoon commented on src/main/java/command/ManageCommand.java:
```diff
@@ -20,17 +20,18 @@ public class ManageCommand extends Command {
+    public ManageCommand(Command.Types type, int... indices) {
         this.type = type;
-        this.index = index;
+        this.indices = indices;
```
Great distinction to separate the use of the word index and the comment! The small things in readability add up 👍

ChuaZenKhoon commented on src/main/java/command/ManageCommand.java:
```diff
@@ -70,33 +71,35 @@ public String getTestData() {
      * @param storage The Storage where changes are saved.
      */
```
Oops, I think you might have forgotten @return tag with the relevant return object?

ChuaZenKhoon commented on src/main/java/parser/Parser.java:
```diff
@@ -126,8 +129,11 @@ public Command unmark(String input) throws LuluException {
         if (words.length <= 1) {
             throw new InvalidCommandException();
```
Uh you might have forgotten to put throws InvalidCommandException in your function descriptor next to your LuluException? Since you don't catch it in this function :) Same for some other functions in this file as well
The javadocs would have another line for that too! 😄

ChuaZenKhoon reviewed (commented):
Great job!

chengjunyuan commented on src/main/java/lulu/Lulu.java:
```diff
@@ -0,0 +1,116 @@
+                }
+            } catch (InvalidCommandException e) {
+                UI.print("Sorry, I don't think I quite understood what you meant...");
+            } catch (InvalidDateException e) {
```
I think it would be good to give an example of a valid date for the user to refer to in your error message! Apart from that, excellent job in your error handling

chengjunyuan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,266 @@
+     *
+     * @param input The user input containing the search query.
+     * @return A {@code QueryCommand} object with the search query and type.
+     * @throws InvalidCommandException If the input command is invalid.
```
Good use of JavaDoc to explain the error handling of the method

chengjunyuan commented on src/main/java/command/ManageCommand.java:
```diff
@@ -0,0 +1,169 @@
+     * Unmarks the task at the specified index, updating its status in the task list and storage.
+     *
+     * @param tasks   The TaskList from which the task is unmarked.
+     * @param storage The Storage where changes are saved.
```
Like the other commenter mentioned above, `@return` param may be missing

chengjunyuan commented on src/main/java/command/ManageCommand.java:
```diff
@@ -0,0 +1,169 @@
+     * Deletes the task at the specified index, removing it from the task list and storage.
+     *
+     * @param tasks   The TaskList from which the task is deleted.
+     * @param storage The Storage where changes are saved.
```
Like the other commenter mentioned above, `@return` param may be missing

chengjunyuan commented on src/main/java/lulu/Lulu.java:
```diff
@@ -0,0 +1,116 @@
+     * Processes user input and generates a response based on the given command.
+     *
+     * @param input The user input to be processed.
+     * @return A response generated based on the user input.
```
Very detailed JavaDoc, well done!

alfaloo committed:
change file path

alfaloo committed:
Enhance code clarity with assert statements

Current: Code lacks explicit documentation.

Why change: Improve code readability and debugging
by assert statements to document critical
assumptions.

What: Integrate assert statements strategically
to show expected conditions at crucial points in
the code.

Why this way: Assert statements provide a
concise and expressive means to document
assumptions effectively.

Other: Focus on critical areas; future commits may extend
assert usage for additional assumptions.

alfaloo committed:
Improve code quality (A-CodeQuality)

Current: Code needs refactoring.
Why: Enhance readability, maintainability.
What: Critical examination, refactoring.
How: Follow 'Code Quality' principles.
Other: Focus on identified improvements.

alfaloo committed:
Utilize Java Streams (A-Streams)

Current: Manual task filtering.
Why: Simplify with Java Streams.
What: Refactor filtering logic.
How: Use `filter` for clarity.
Other: Focus on task filtering.

alfaloo committed:
Merge pull request #1 from alfaloo/branch-A-Assertions

Enhance code clarity with assert statements

alfaloo committed:
Merge pull request #2 from alfaloo/branch-A-CodeQuality

Improve code quality (A-CodeQuality)

alfaloo committed:
Merge pull request #3 from alfaloo/branch-A-Streams

Utilize Java Streams (A-Streams)

alfaloo committed:
add extension

alfaloo committed:
Use GitHub Actions to set up Continuous Integration

alfaloo committed:
minor changes

alfaloo committed:
fix bugs

alfaloo committed:
fix bugs

alfaloo committed:
deal with duplicates with hashset instead of arraylist

alfaloo committed:
update javadocs

alfaloo committed:
add more error handling

alfaloo committed:
add more junit tests

alfaloo committed:
update user guide

alfaloo committed:
improve user guide

alfaloo committed:
change file path generation

alfaloo committed:
update file path generation

alfaloo committed:
improve coding standard

alfaloo committed:
update personality

alfaloo committed:
document ai usage

alfaloo committed:
update personality

alfaloo committed:
update stylistic issues

alfaloo committed:
fix minor bugs

alfaloo committed:
update and improve user guide

alfaloo committed:
update and improve readme

alfaloo committed:
update minor issues

alfaloo committed:
update user guide to include installation instructions

alfaloo committed:
update minor stylistic issues
